### PR TITLE
Game Mod Descriptors have their own INI configuration

### PR DESCRIPTION
**Pull request by Zarklord**
This add support for Game Mod Descriptors (the json files currently only used for riivolution mods) to have their own config INI.
This INI uses the display name of the mod to have a seperate INI file for configuration.

The uses cases for something like this is having a set of gecko codes, ar codes, patches, ect that you want all enabled as on, as part of some modpack or something, but would still like to be able to play the original game without having to go configure gecko codes and such every time you launch the game.

Due to how these work, these JSON mods are fully compatible for netplay, since it still launches the underlying game, with just a custom INI.

This images shows what a user see's in the editor view(note the `Legacy XP.ini`)
image

Currently you can only manually create these json configuration preset mods, but I will work on another PR in the future that adds support for creating custom Configuration Presets from the UI itself.

### CODE REVIEW

File: Source/Core/Core/ConfigManager.cpp [hunk 1]
```diff
@@ -374,6 +373,10 @@ bool SConfig::SetPathsAndGameMetadata(const BootParameters& boot)
   if (!std::visit(SetGameMetadata(this, &m_region), boot.parameters))
     return false;
 
+  m_local_config_override = boot.config_override;
+  Config::AddLayer(
+      ConfigLoaders::GenerateLocalGameConfigLoader(m_local_config_override, m_revision));
+
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/ConfigManager.cpp:379]: This looks fishy... wouldn't you want to use the game ID in the not-overriden case?

Comment (AdmiralCurtiss) [reply Source/Core/Core/ConfigManager.cpp:379]: You probably also should move the Config::AddLayer() back to where it was before -- at the very least, it needs to be called before the `DolphinAnalytics::Instance().ReportGameStart();` call.

**Comment by AdmiralCurtiss** [comment]
I think allowing a separate user GameINI for mod descriptor files is sensible. However, I would recommend instead making a new field in the JSON that specifies where that INI is, instead of tying it to the display name.

File: Source/Core/Core/ConfigManager.cpp (hunk 1 above)
Comment (Zarklord) [reply Source/Core/Core/ConfigManager.cpp:379]: Whoops, good catch that should be a call to GetLocalConfig()

**Comment by Zarklord** [comment]
The problem with that, is the issue of how you would present this to a user creating one of these(when we have UI for that), they are already naming it, and now we are asking for a name for the configuration file(which for an end user is fairly irrelevant, as long as it just works), and that would probably just cause confusion for a user trying create these game mod presets.

I think maybe a better solution would be to have a property in the json that if present is used instead of the Display Name for the INI file, this would allow power users to have that level of control that they might want, but it remains clean and simple for everyone else.

**Comment by AdmiralCurtiss** [comment]
Just make it a path -- like the path to the base ISO, relative to the position of the json file -- and default it to the same filename as the json file itself (but with an .ini extension instead of .json) when you create a new 'GameINI' for this in the GUI. That puts the ini right next to the json file in the same directory, simple and straightforward, and if someone wants to do something fancy they still can.

**Comment by AdmiralCurtiss** [comment]
The reason I'm against using the display name, for reference, is that that may contain an arbitrary string that may not even be resolvable in the current file system. For example if it contains a `:` on Windows.

**Comment by Zarklord** [comment]
Making it a path, poses a whole set of problems, and requires re-writing almost the code in the pull request to support this (since currently these configs are still stored in the user config folder), I think sanitizing Display Name is probably a much better option to solve that particular issue.

**Comment by Pokechu22** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
If you want to go with the display name approach, I would suggest prepending "ID-" to the game INI name, to match how we handle DOL and ELF files.

Also, am I misreading the code or does this break loading the base game INI for game mod descriptors? That could be annoying, depending on how you're using game INIs.

**Comment by Zarklord** [comment]
@JosJuice that's a good idea, I will go append "ID-" to the name.

When you say break loading the base game INI, are you talking about the user customized one stored in the user directory, or the defaults that get shipped with dolphin?
If its the latter then no they load just fine, if its the former then yes, that would break, but I would (in general) not expect anyone currently using game mod descriptors would already have to either:
- go manually enable/disable a bunch of things when loading riivolution mods vs the base game
- or they have all default settings, and need to do nothing launching these titles

In the case of of the first one they would have more work once, and then less work every time after, and the second one nothing would be noticeably different.

**Comment by Zarklord** [comment]
Okay, so "ID-" is now prepended onto the Game Mod Descriptor name.

Additionally the display name is no longer used for the ini name, instead just like with elf/dol files, the file name of the json is used, so if you had a `Brawl Minus.json` it would use the INI `ID-Brawl Minus.ini`.

An additional option in the json `config-ini-override` has been added, if its there it will use that name instead of the file name, so for the previous example if you had `config-ini-override` equal to "BrawlMinusNetplay" it would instead use `ID-BrawlMinusNetplay.ini` instead of `ID-Brawl Minus.ini`

**Comment by Pokechu22** [comment]
@dolphin-emu-bot rebuild

**Comment by Pokechu22** [comment]
@dolphin-emu-bot rebuild

**Comment by Pokechu22** [comment]
Dolphin actually supports multiple local configs in the user directory, with the same filename patterns as those shipped with Dolphin.  Although creating them isn't exposed in Dolphin's UI, you _can_ manually create files in the user directory: a `RSB.ini` that applies to all languages of Smash Bros Brawl, or a `RSBE01r1.ini` and `RSBE01r12ini` that apply to revisions 1 and 2 specifically (and then those files will appear in the UI and can be edited normally).

It seems like none of the extra local configs are loaded, when before they were loaded for game mod descriptors.  Similarly, a file such as `ID-Super Smash Test.ini` in the global configuration folder isn't loaded (such a file probably won't ever be shipped, but it's something that theoretically could work).

The problem of creating extra configs not being exposed in Dolphin's UI does have a simple solution: always show the ID for the specific game being edited (e.g. when editing the base Smash Bros Brawl, always show `RSBE01.ini` even if `RSB.ini` or `RSBE01r1.ini` exists, and when editing a mod always show the INI for the mod even if parent ones exist).  This isn't the current behavior, though (if one INI exists in the local folder, then that one will be shown with no spot to create a new one), and we would only want that behavior to apply to the user config and not the default config.

**Comment by Zarklord** [comment]
> Dolphin actually supports multiple local configs in the user directory, with the same filename patterns as those shipped with Dolphin. Although creating them isn't exposed in Dolphin's UI, you _can_ manually create files in the user directory: a `RSB.ini` that applies to all languages of Smash Bros Brawl, or a `RSBE01r1.ini` and `RSBE01r12ini` that apply to revisions 1 and 2 specifically (and then those files will appear in the UI and can be edited normally).

Yes, your right, but those local configs still all get applied(or at least get a applied in a non helpful way in relation to this PR)
I can't create a preset for a game mod (like Project M, Brawl Minus, ect) with the current system, its useful to have the current system, but it can't allow you to launch a game with a specified set of configs, and then immediately launch the vanilla game, currently you would still need to go and uncheck gecko codes, changes settings ect.

> It seems like none of the extra local configs are loaded, when before they were loaded for game mod descriptors. 

This should still work, it would just be `ID-Super Smash Testr1.ini` or `ID-Super Smash Testr2.ini`

> A file such as `ID-Super Smash Test.ini` in the global configuration folder isn't loaded (such a file probably won't ever be shipped, but it's something that theoretically could work).

Unless we cross into shipping custom mod descriptor presets, we don't need them from the global folder, since these would only exist for user created mod descriptors(which have will arbitrary names), I don't even think it makes sense to implement something like that.

> The problem of creating extra configs not being exposed in Dolphin's UI does have a simple solution: always show the ID for the specific game being edited (e.g. when editing the base Smash Bros Brawl, always show `RSBE01.ini` even if `RSB.ini` or `RSBE01r1.ini` exists, and when editing a mod always show the INI for the mod even if parent ones exist). This isn't the current behavior, though (if one INI exists in the local folder, then that one will be shown with no spot to create a new one), and we would only want that behavior to apply to the user config and not the default config.

I'm not quite sure I'm understanding this, could you rephrase this or something, its just not making any sense in my head.

**Comment by Zarklord** [comment]
@AdmiralCurtiss sorry I have been busy and forgot to ask, is there something that I need to do on my end to potentially get this merged? its somewhat annoying having to maintain my own dolphin fork with my changes, and would really appreciate if this could get merged, or if I can get informed of what I need to fix to get it merged, thanks.

**Comment by Avasam** [comment]
If I understand this correctly, it would be useful for Aspect Ratio hacks that could also set the correct Custom Aspect ratio in Dolphin settings.

> is there something that I need to do on my end to potentially get this merged?

I'm not a maintainer, but your branch has conflicts now, so at the minimum there's that.

**Comment by Zarklord** [comment]
> If I understand this correctly, it would be useful for Aspect Ratio hacks that could also set the correct Custom Aspect ratio in Dolphin settings.
> 
> > is there something that I need to do on my end to potentially get this merged?
> 
> I'm not a maintainer, but your branch has conflicts now, so at the minimum there's that.

I'd be happy to fixup the PR if I got confirmation that this will be merged (after reviewing + any needed changes), but otherwise I don't have the desire to touch this further.

**Comment by Anuskuss** [comment]
This doesn't work when the game is launched from the command line.

**Comment by Zarklord** [comment]
> This doesn't work when the game is launched from the command line.

I would be shocked if this worked at all at this point, this PR is over 2 years old now...

My offer still stands to spend time to get this working again, provided it will actually be merged into Dolphin, otherwise I won't waste my time maintaining this fork.
